--- a/net/minecraft/server/Main.java
+++ b/net/minecraft/server/Main.java
@@ -5,6 +_,8 @@
 import com.mojang.logging.LogUtils;
 import com.mojang.serialization.Dynamic;
 import com.mojang.serialization.Lifecycle;
+import dev.akarah.faarah.server.plugin.PluginLoader;
+import dev.akarah.faarah.server.plugin.PluginRepository;
 import java.awt.GraphicsEnvironment;
 import java.io.File;
 import java.io.IOException;
@@ -69,6 +_,13 @@
     )
     @DontObfuscate
     public static void main(String[] args) {
+        // Faarah start - Load plugins during startup
+        var pluginLoader = PluginLoader.of();
+        pluginLoader.loadFromDirectories(Path.of("./plugins/"));
+        pluginLoader.load();
+        PluginRepository.getInstance().startupAllPlugins();
+        // Faarah end
+
         SharedConstants.tryDetectVersion();
         OptionParser optionParser = new OptionParser();
         OptionSpec<Void> optionSpec = optionParser.accepts("nogui");
